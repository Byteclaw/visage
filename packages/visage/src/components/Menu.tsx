import {
  ExtractVisageComponentProps,
  OmittableProps,
  markAsVisageComponent,
} from '@byteclaw/visage-core';
import React, {
  Children,
  cloneElement,
  useCallback,
  useEffect,
  KeyboardEventHandler,
  forwardRef,
  RefObject,
  Ref,
} from 'react';
import {
  findNextMatchingSiblingElement,
  findPreviousMatchingSiblingElement,
  PlacementWithAnchorOrigin,
  Placement,
} from './shared';
import { createComponent } from '../core';
import { useCombinedRef } from '../hooks';
import { List, ListItem } from './List';
import { Popover } from './Popover';

const MenuBase = createComponent(List, {
  displayName: 'Menu',
  styles: {
    backgroundColor: 'shades',
  },
});

const MenuItemBase = createComponent(ListItem, {
  displayName: 'MenuItem',
  styles: {
    '&:hover, &:focus, &[aria-selected="true"]': {
      outline: 'none',
      backgroundColor: 'primary.1',
      color: 'primaryText.1',
      cursor: 'pointer',
      userSelect: 'none',
    },
  },
});

interface MenuProps extends ExtractVisageComponentProps<typeof MenuBase> {
  anchor?: null | RefObject<HTMLElement>;
  /**
   * A ref to base HTML element
   */
  baseRef?: RefObject<HTMLDivElement>;
  /**
   * Use only if you are managing focus outside of this component
   */
  disableAutoFocusItem?: boolean;
  /**
   * Disable all default event handlers on menu?
   *
   * The purpose of this props is for example if you want to use Menu as companion to TextInput
   */
  disableEvents?: boolean;
  /**
   * Make menu as long as anchor element?
   */
  keepAnchorWidth?: boolean;
  onClose?: () => void;
  /**
   * Is menu open?
   */
  open?: boolean;
  /**
   * A responsive array of prioritized placements
   *
   * Can be used to say how should popover behave in different scenarios
   *
   * Default is top left placement with top left anchor origin and bottom left alternative
   */
  placement?: (PlacementWithAnchorOrigin[] | undefined)[];
  /**
   * Props that will be passed to underlying Popover component
   */
  popoverProps?: OmittableProps<ExtractVisageComponentProps<typeof Popover>>;
}

type MenuItemProps = ExtractVisageComponentProps<typeof MenuItemBase>;

const defaultPlacement: PlacementWithAnchorOrigin[][] = [
  [
    { placement: Placement.topLeft, vertical: 'bottom', horizontal: 'left' },
    {
      placement: Placement.bottomLeft,
      vertical: 'top',
      horizontal: 'left',
    },
  ],
];

const popoverStyles = {
  boxShadow: '0 0 0 1px rgba(63,63,68,.05), 0 1px 3px 0 rgba(63,63,68,.60)',
};

const isMenuItemElement = (el: HTMLElement) =>
  el.getAttribute('role') === 'menuitem';

export function Menu({
  anchor,
  baseRef,
  children,
  disableAutoFocusItem,
  disableEvents,
  keepAnchorWidth,
  onClose,
  open = false,
  placement = defaultPlacement,
  role = 'menu',
  onKeyDown: outerOnKeyDown,
  popoverProps,
  ...restProps
}: MenuProps) {
  const menuRef = useCombinedRef(baseRef);
  const onKeyDown: KeyboardEventHandler<HTMLElement> = useCallback(
    e => {
      if (outerOnKeyDown) {
        outerOnKeyDown(e);
      }

      if (disableEvents) {
        return;
      }

      switch (e.key) {
        case 'ArrowUp': {
          // prevent so Popover onKeyDown is not called
          e.stopPropagation();
          e.preventDefault();
          // find previous focusable element, if none is found, find the last one
          const previousFocusableElement = findPreviousMatchingSiblingElement(
            e.currentTarget as HTMLElement,
            isMenuItemElement,
          );

          if (previousFocusableElement) {
            previousFocusableElement.focus();
            return;
          }
          if (menuRef.current) {
            // focus last
            menuRef.current
              .querySelector<HTMLElement>('[role="menuitem"]:last-of-type')
              ?.focus();
          }

          return;
        }
        case 'ArrowDown': {
          // prevent so Popover onKeyDown is not called
          e.stopPropagation();
          e.preventDefault();
          // find next focusable element, if none is found, find the first one
          const nextFocusableElement = findNextMatchingSiblingElement(
            e.currentTarget as HTMLElement,
            isMenuItemElement,
          );

          if (nextFocusableElement) {
            nextFocusableElement.focus();
          } else if (menuRef.current) {
            // focus first
            menuRef.current
              .querySelector<HTMLElement>('[role="menuitem"]:first-of-type')
              ?.focus();
          }
        }
      }
    },
    [disableEvents, outerOnKeyDown],
  );

  const onPopoverKeyDown: KeyboardEventHandler<HTMLElement> = useCallback(
    e => {
      if (disableEvents) {
        return;
      }

      switch (e.key) {
        case 'ArrowUp': {
          e.preventDefault();

          // focus last menu item
          if (menuRef.current) {
            menuRef.current
              .querySelector<HTMLElement>('[role="menuitem"]:last-of-type')
              ?.focus();
          }

          return;
        }
        case 'ArrowDown': {
          e.preventDefault();

          // focus first menu item
          if (menuRef.current) {
            menuRef.current
              .querySelector<HTMLElement>('[role="menuitem"]:first-of-type')
              ?.focus();
          }
        }
      }
    },
    [disableEvents],
  );

  // manage autofocus of first item
  // if not managed from outside
  useEffect(() => {
    if (!open || disableAutoFocusItem || disableEvents) {
      return;
    }

    if (menuRef.current) {
      const firstMenuItem = menuRef.current.querySelector<HTMLElement>(
        '[role="menuitem"]:first-of-type',
      );

      if (firstMenuItem) {
        firstMenuItem.focus();
      }
    }
  }, [open, disableAutoFocusItem, disableEvents]);

  return (
    <Popover
      anchor={anchor}
      autoFocus={!disableEvents && disableAutoFocusItem}
      backdrop
      keepAnchorWidth={keepAnchorWidth}
      onClose={onClose}
      onKeyDown={onPopoverKeyDown}
      open={open}
      placement={placement}
      minHeight={150}
      {...popoverProps}
      styles={{
        ...popoverStyles,
        ...popoverProps?.styles,
      }}
    >
      <MenuBase ref={menuRef} role={role} {...restProps} tabIndex={-1}>
        {Children.map(children, menuItem => {
          return cloneElement(menuItem as any, {
            onKeyDown,
          });
        })}
      </MenuBase>
    </Popover>
  );
}

export const MenuItem: typeof MenuItemBase = forwardRef(
  (
    { children, role = 'menuitem', ...restProps }: MenuItemProps,
    ref: Ref<HTMLLIElement>,
  ) => {
    // https://www.w3.org/TR/wai-aria-practices-1.1/#wai-aria-roles-states-and-properties-13
    return (
      <MenuItemBase tabIndex={-1} role={role} ref={ref} {...restProps}>
        {children}
      </MenuItemBase>
    );
  },
) as any;

MenuItem.displayName = 'MenuItem';

markAsVisageComponent(MenuItem);
