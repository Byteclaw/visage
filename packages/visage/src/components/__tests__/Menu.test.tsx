import { fireEvent } from '@testing-library/react';
import React from 'react';
import { render } from './render';
import { Menu, MenuItem } from '../Menu';

describe('Menu', () => {
  it('automatically focuses first item on open and navigates using arrows', () => {
    const onFocus = jest.fn();
    const anchor = document.createElement('div');
    document.body.append(anchor);

    render(
      <Menu anchor={anchor} open>
        <MenuItem onFocus={() => onFocus('A')}>A</MenuItem>
        <MenuItem onFocus={() => onFocus('B')}>B</MenuItem>
        <div />
        <MenuItem onFocus={() => onFocus('C')}>C</MenuItem>
      </Menu>,
    );

    // expect first to be focused
    expect(onFocus).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledWith('A');

    // now push arrow down to select B
    fireEvent.keyDown(
      document.querySelector('[role="menuitem"]:nth-child(1)'),
      { key: 'ArrowDown' },
    );

    expect(onFocus).toHaveBeenCalledTimes(2);
    expect(onFocus).toHaveBeenCalledWith('B');

    // select C
    fireEvent.keyDown(
      document.querySelector('[role="menuitem"]:nth-child(2)'),
      { key: 'ArrowDown' },
    );

    expect(onFocus).toHaveBeenCalledTimes(3);
    expect(onFocus).toHaveBeenCalledWith('C');

    // arrow down should select A again
    fireEvent.keyDown(
      document.querySelector('[role="menuitem"]:nth-child(4)'),
      { key: 'ArrowDown' },
    );

    expect(onFocus).toHaveBeenCalledTimes(4);
    expect(onFocus).toHaveBeenCalledWith('A');

    // arrow up should select C again
    fireEvent.keyDown(
      document.querySelector('[role="menuitem"]:nth-child(1)'),
      { key: 'ArrowUp' },
    );

    expect(onFocus).toHaveBeenCalledTimes(5);
    expect(onFocus).toHaveBeenCalledWith('C');
  });

  it('does not autofocus first item if disableAutoFocusItem is provided', () => {
    const onFocus = jest.fn();
    const anchor = document.createElement('div');
    document.body.append(anchor);

    render(
      <Menu anchor={anchor} disableAutoFocusItem open>
        <MenuItem onFocus={() => onFocus('A')}>A</MenuItem>
        <MenuItem onFocus={() => onFocus('B')}>B</MenuItem>
        <div />
        <MenuItem onFocus={() => onFocus('C')}>C</MenuItem>
      </Menu>,
    );

    expect(onFocus).not.toHaveBeenCalled();

    // arrow down selects A
    fireEvent.keyDown(document.querySelector('[role="menu"]'), {
      key: 'ArrowDown',
    });

    expect(onFocus).toBeCalledTimes(1);
    expect(onFocus).toBeCalledWith('A');
  });

  it('does not react to keyboard events if disableEvents is provided', () => {
    const onFocus = jest.fn();
    const anchor = document.createElement('div');
    document.body.append(anchor);

    render(
      <Menu anchor={anchor} disableEvents open>
        <MenuItem onFocus={() => onFocus('A')}>A</MenuItem>
        <MenuItem onFocus={() => onFocus('B')}>B</MenuItem>
        <div />
        <MenuItem onFocus={() => onFocus('C')}>C</MenuItem>
      </Menu>,
    );

    expect(onFocus).not.toHaveBeenCalled();

    fireEvent.keyDown(document.querySelector('[role="menu"]'), {
      key: 'ArrowDown',
    });

    expect(onFocus).not.toBeCalled();
  });
});
