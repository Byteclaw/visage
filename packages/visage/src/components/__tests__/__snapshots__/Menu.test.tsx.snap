// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu is possible to extend menu 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 0;
  height: auto;
  margin: 0px;
  overflow: hidden;
  width: auto;
}

.emotion-0[role="button"]:hover,
.emotion-0[role="button"]:focus {
  outline: none;
  background-color: rgb(0,0,255);
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0[data-gutters="true"] {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-0[data-active="true"] {
  color: rgb(173,216,230);
  font-weight: bolder;
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0[aria-selected="true"] {
  outline: none;
  background-color: rgb(0,0,255);
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-1 {
  display: block;
  list-style: none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  margin: 0px;
  padding: 0px;
  padding-top: 16px;
  padding-bottom: 16px;
  width: auto;
  background-color: rgb(204,204,204);
}

.emotion-1[data-collapsed="true"] {
  max-height: 0;
  visibility: hidden;
  padding: 0px;
  margin: 0px;
}

<ul
  class="emotion-1"
  data-collapsed="false"
  data-testid="menu"
  role="menu"
  tabindex="-1"
>
  <li
    class="emotion-0"
    data-active="false"
    data-button="false"
    data-gutters="true"
    role="menuitem"
    tabindex="-1"
  >
    A
  </li>
</ul>
`;

exports[`Menu is possible to style menu base using styles prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 0;
  height: auto;
  margin: 0px;
  overflow: hidden;
  width: auto;
}

.emotion-0[role="button"]:hover,
.emotion-0[role="button"]:focus {
  outline: none;
  background-color: rgb(0,0,255);
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0[data-gutters="true"] {
  padding-left: 16px;
  padding-right: 16px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.emotion-0[data-active="true"] {
  color: rgb(173,216,230);
  font-weight: bolder;
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0[aria-selected="true"] {
  outline: none;
  background-color: rgb(0,0,255);
  color: rgb(255,255,255);
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-1 {
  display: block;
  list-style: none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  margin: 0px;
  padding: 0px;
  padding-top: 16px;
  padding-bottom: 16px;
  width: auto;
  background-color: rgb(204,204,204);
}

.emotion-1[data-collapsed="true"] {
  max-height: 0;
  visibility: hidden;
  padding: 0px;
  margin: 0px;
}

<ul
  class="emotion-1"
  data-collapsed="false"
  data-testid="menu"
  role="menu"
  tabindex="-1"
>
  <li
    class="emotion-0"
    data-active="false"
    data-button="false"
    data-gutters="true"
    role="menuitem"
    tabindex="-1"
  >
    A
  </li>
</ul>
`;
